File: genetic.ts
class Chromosome {
    private string: string[];
    private score: number;

    constructor(string: string[], score: number) {
        this.string = string;
        this.score = score;
    }

    public getScore(): number {
        return this.score;
    }

    public getString(): string[] {
        return this.string;
    }

    public setCharacter(index: number, character: string): void {
        this.string[index] = character;
    }
}

class Main {
    readonly low: number = 32;
    readonly high: number = 126;
    readonly POP_SIZE: number = 100;
    private chromosomes: Chromosome[] = [];
    private newChromosomes: Chromosome[] = [];
    private startHighest: number = 0;
    private startHighestString: string[];

    private iteration: number = 0;
    readonly helloWorld: number[] = [72, 101, 108, 108, 111, 44, 32, 87, 111, 114, 108, 100, 33];

    public run() {
        this.fillArray();
        while (!this.checkSolution()) {
            for (let i = 0; i < this.POP_SIZE; i++) {
                this.createNewPopulation();
            }
            let currentHighest = 0;
            let highestString = '';
            for (let i = 0; i < this.POP_SIZE; i++) {
                let score = this.chromosomes[i].getScore();
                if (i == 0 || score < currentHighest) {
                    currentHighest = score;
                    highestString = this.chromosomes[i].getString().join('');
                }
            }
            if (!this.checkSolution()) {
                this.chromosomes = this.newChromosomes.slice();
                this.mutate();
                console.log(`${this.iteration}. Best = ${highestString}, ${currentHighest} off target`)
        
                this.newChromosomes = [];
                this.iteration++;
            }
        }
        console.log('\nHello, World! has been made');
        console.log(`Start string was ${this.startHighestString.join('')} (${this.startHighest} away from target)\n`);
    }

    private createString(): string[] {
        let string: string[] = [];
        for (let i = 0; i < 13; i++) {
            let letter: number = this.random(this.low, this.high);
            string.push(String.fromCharCode(letter));
        }
        return string;
    }

    private fillArray(): void {
        let string, score;
        for (let i = 0; i < this.POP_SIZE; i++) {
            string = this.createString();
            this.chromosomes.push(new Chromosome(string, this.calculateFitness(string)));
            score = this.calculateFitness(string);
            if (i == 0 || score < this.startHighest) {
                this.startHighest = score;
                this.startHighestString = this.chromosomes[i].getString();
            }
        }
    }

    private calculateFitness(string): number {
        let score = 0;
        for (let i = 0; i < string.length; i++) {
            score += Math.abs(string[i].charCodeAt(0) - this.helloWorld[i]);
        }
        return score;
    }

    private constructString(arr): string {
        let string = '';
        arr.forEach(char => {
            string = string + char;
        });
        return string;
    }

    private checkSolution(): boolean {
        for (let i =0; i < this.POP_SIZE; i++) {
            if (this.chromosomes[i].getScore() === 0) {
                return true;
            }
        }
        return false;
    }

    private crossOver(parent1, parent2): void {
        let rand = this.random(parent1.length);
        let child1 = '';
        let child2 = '';

        for (let i = 0; i < parent1.length; i++) {
            child1 += i >= rand ? parent1[i] : parent2[i];
            child2 += i >= rand ? parent2[i] : parent1[i];
        }

        this.newChromosomes.push(new Chromosome(child1.split(''), this.calculateFitness(child1.split(''))));
        this.newChromosomes.push(new Chromosome(child2.split(''), this.calculateFitness(child2.split(''))));
    }

    private createNewPopulation(): void {
        let randomFour1 = [];
        let randomFour2 = [];

        for (let i = 0; i < 4; i++) {
            randomFour1.push(this.random(this.chromosomes));
            randomFour2.push(this.random(this.chromosomes));
        }

        let parent1: string[] = this.fitnessFunction(randomFour1).getString();
        let parent2: string[] = this.fitnessFunction(randomFour2).getString();
        let n = this.random(1, 100);
        if (n < 70) {
            this.crossOver(parent1, parent2);
        } else {
            this.newChromosomes.push(new Chromosome(parent1, this.calculateFitness(parent1)));
            this.newChromosomes.push(new Chromosome(parent2, this.calculateFitness(parent2)));
        }
    }

    private mutate(): void {
        for (let i = 0; i < this.chromosomes.length; i++) {
            if (this.random(1, 100) < 5) {
                let StringToMutate = this.chromosomes[i].getString();
                let answer = this.random(StringToMutate.length);
                let chromosomeString = this.chromosomes[i].getString()[answer];
                let character = chromosomeString.charCodeAt(0);
                if (character == this.low) {
                    character += 1;
                } else if (character == this.high) {
                    character -= 1;
                } else {
                    this.random(2) == 0 ? character += 1 : character -= 1;
                }
                this.chromosomes[i].setCharacter(answer, String.fromCharCode(character));
            }
        }
    }

    private fitnessFunction(randomFour): Chromosome {
        let fitnessScore = 0;
        let winnerPosition = 0;
        for (let i = 0; i < randomFour.length; i++) {
            if (i == 0 || randomFour[i].score < fitnessScore) {
                winnerPosition = i;
                fitnessScore = randomFour[i].getScore();
            }
        }
        return randomFour[winnerPosition];
    }

    private random(x: any, y?: number): any {
        if (typeof y === 'undefined') {
            if (x instanceof Array) {
                let index: number = Math.floor(Math.random() * x.length);    
                return x[index];
            } else {
                return Math.floor(Math.random() * x);            
            }
        } else {
            return Math.floor(Math.random() * (y - x) + x);
        }
    }
}
let mainMethod = new Main();

mainMethod.run();

File: genetic.kt
import java.util.Random

val helloWorld: IntArray = intArrayOf(72, 101, 108, 108, 111, 44, 32, 87, 111, 114, 108, 100, 33)
var HIGH: Int = 126
var LOW: Int = 32
var POP_SIZE: Int = 100
var startHighest = 0
var startHighestString = ""
var chromosomes: MutableList<Chromosome> = mutableListOf()
var newChromosomes: MutableList<Chromosome> = mutableListOf()

fun main(args: Array<String>) {
    fillArray()
    var iteration = 0
    while (!checkSolution()) {
        var currentHighest = 0
        var highestString = ""
        for(i in 0 until POP_SIZE) {
            if (i < POP_SIZE/2) {
            createNewPopulation()
            }
            if(i == 0 || chromosomes[i].score < currentHighest) {
                currentHighest = chromosomes[i].score
                highestString = chromosomes[i].string
            }
        }
        if(!checkSolution()) {
            chromosomes = newChromosomes.toMutableList()
            mutate()
            println("$iteration. Best = $highestString, $currentHighest off target")
            newChromosomes = mutableListOf()
            iteration++
        }
    }
    println("\nHello, World! has been made")
    println("Start string was $startHighestString ($startHighest away from target)")
}

class Chromosome(var string: String, var score: Int) {

}

fun createString(): String {
    val sb = StringBuilder()
    val rand = Random()
    for (i in 0 until 13) {
        var n = rand.nextInt((HIGH - LOW) + 1) + LOW
        sb.append(n.toChar())
    }
    return sb.toString()
}

fun calculateFitness(string: String): Int {
    var score = 0
    val stringToCheck = string.toCharArray()
    for (i in 0 until string.length) {
        score += Math.abs((stringToCheck[i].toInt()) - helloWorld[i])
    }
    return score
}

fun fillArray() {
    var word = ""
    var score = 0
    var chromosome: Chromosome
    for (i in 0 until POP_SIZE) {
        word = createString()
        score = calculateFitness(word)
        chromosome = Chromosome(word, score)
        chromosomes.add(chromosome)
        if(i == 0 || score < startHighest) {
            startHighest = score
            startHighestString = word
        }
    }
}

fun checkSolution(): Boolean {
    for(chromosome in chromosomes) {
        if(chromosome.score == 0) {
            return true
        }
    }
    return false
}

fun createNewPopulation() {
    var randomFour1: MutableList<Chromosome> =  mutableListOf()
    var randomFour2: MutableList<Chromosome> = mutableListOf()
    val rand = Random()

    for(i in 1..4) {
        randomFour1.add(chromosomes[rand.nextInt(POP_SIZE)])
        randomFour2.add(chromosomes[rand.nextInt(POP_SIZE)])
    }

    val parent1 = fitnessFunction(randomFour1).string
    val parent2 = fitnessFunction(randomFour2).string
    val n = rand.nextInt(100 - 1 + 1) + 1
    if(n < 70) {
        crossOver(parent1, parent2)
    } else {
        newChromosomes.add(Chromosome(parent1, calculateFitness(parent1)))
        newChromosomes.add(Chromosome(parent2, calculateFitness(parent2)))
    }
}

fun crossOver(parent1: String, parent2: String) {
    val rand = Random()
    val position = rand.nextInt(parent1.length)
    val child1Chars = parent1.toCharArray()
    val child2Chars = parent2.toCharArray()
    var child1 = StringBuilder()
    var child2 = StringBuilder()
    for(i in 0 until parent1.length) {
        child1.append(if (i >= position) child1Chars[i] else child2Chars[i])
        child2.append(if (i >= position) child2Chars[i] else child1Chars[i])
    }
    newChromosomes.add(Chromosome(child1.toString(), calculateFitness(child1.toString())))
    newChromosomes.add(Chromosome(child2.toString(), calculateFitness(child2.toString())))

}

fun fitnessFunction(randomFour: MutableList<Chromosome>): Chromosome {
    var fitnessScore = 0
    var winnerPosition = 0
    for (i in 0 until randomFour.size) {
        var winnerValue = 0
        var word = randomFour[i].string.toCharArray()
        for(j in 0 until word.size) {
            winnerValue +=  Math.abs((word[j].toInt()) - helloWorld[j])
        }
        if(i == 0 || winnerValue < fitnessScore) {
            winnerPosition = i
            fitnessScore = winnerValue
        }
    }
    return randomFour[winnerPosition]
}

fun mutate() {
    val rand = Random()
    for (chromosome in chromosomes) {
        val n = rand.nextInt((100 - 1) + 1) + 1
        if (n < 5) {
            var stringToMutate = chromosome.string
            var answer = rand.nextInt(stringToMutate.length)
            var wordChar = stringToMutate.toCharArray()
            var changedCharacter: Int
            if (wordChar[answer].toInt() == 32) {
                changedCharacter = wordChar[answer].toInt() + 1
            }
            else if (wordChar[answer].toInt() == 126) {
                changedCharacter = wordChar[answer].toInt() - 1
            }
            else {
                changedCharacter = if (rand.nextInt(2) == 0)  wordChar[answer].toInt() + 1 else wordChar[answer].toInt() - 1
            }

            wordChar[answer] = changedCharacter.toChar()
            var mutatedString = ""
            for (i in 0 until wordChar.size) {
                mutatedString += wordChar[i]
            }
            chromosome.string = mutatedString
            chromosome.score = calculateFitness(mutatedString)
        }
    }
}